src/mon/OSDMonitor.cc

```diff
@@ -7954,6 +7954,22 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
 	ss << "create-or-move updated item name '" << osd_name
 	   << "' weight " << weight
 	   << " at location " << loc << " to crush map";
+
+        // If the OSD is destroyed use initial_weight to adjust the weight of the
+        // item in the CRUSHMap as the underlying device might be a different size
+        // For example a 4TB device was replaced by a 6TB device, the CRUSH weight
+        // then needs to be adjusted accordingly
+        //
+        // The OSD will be in the 'new' state at that moment as it has already been
+        // booted and the 'destroyed' state has been removed from the OSD.
+        // At this point it is marked as 'exists,new'
+        if (osdmap.is_new(osdid)) {
```

Comment:
@wido i think we should update `CrushWrapper::create_or_move_item()` instead, because in the case of "move", the weight is not specified, and we need to pass the original weight to the underlying `CrushWrapper::insert_item()`, that's why we overwrites the `weight` passed to `CrushWrapper::create_or_move_item()` with the one returned by `get_item_weightf(item)`.

i think, instead of overwriting the `weight` we could pass `-1` to `create_or_move_item()` when we intend to "move" an item to indicate that we need preserve the original weight of this item instead of using the parameter.

something like:
```c++
    if (weight < 0 && _search_item_exists(item)) {
      weight = get_item_weightf(item);
      ldout(cct, 10) << "create_or_move_item " << item
		     << " exists with weight " << weight << dendl;
      remove_item(cct, item, true);
    }
```

and update the caller of `CrushWrapper::create_or_move_item()` in OSDMonitor.cc and crushtools.cc.

what do you think?